execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -30,26 +30,33 @@ use crate::{
 type SharedCache = Arc<RwLock<Cache>>;
 
 struct Cache {
-    stored_values: HashMap<Key, StoredValue>,
+    cached_values: HashMap<Key, (bool, StoredValue)>,
```

Comment:
Alternatively, we could have separate cache for read-only and never write those back. I think something similar was done for `TrackingCopy`.